mne/morph.py

```diff
@@ -0,0 +1,1477 @@
+
+        elif self.kind == 'surface' or self.kind == 'vector':
+            s += ", smooth : %s" % self.smooth
+            s += ", xhemi : %s" % self.xhemi
+
+        return "<SourceMorph  |  %s>" % s
+
+    @verbose
+    def save(self, fname, verbose=None):
+        """Save the morph for source estimates to a file.
+
+        Parameters
+        ----------
+        fname : str
+            The stem of the file name. '-morph.h5' will be added if fname does
+            not end with '.h5'
+        verbose : bool | str | int | None
+            If not None, override default verbose level (see
+            :func:`mne.verbose` and :ref:`Logging documentation <tut_logging>`
+            for more).
+        """
+        logger.info('saving morph...')
+        if not fname.endswith('.h5'):
+            fname = '%s-morph.h5' % fname
+
+        write_hdf5(fname, self.__dict__, overwrite=True)
+        logger.info('[done]')
+
+    def as_volume(self, stc, fname=None, mri_resolution=False, mri_space=True,
+                  apply_morph=False, format='nifti1'):
+        """Return volume source space as Nifti1Image and / or save to disk.
+
+        Parameters
+        ----------
+        stc : VolSourceEstimate
+            Data to be transformed
+        fname : str | None
+            String to where to save the volume. If not None that volume will
+            be saved at fname.
+        mri_resolution: bool | tuple | int | float
+            Whether to use MRI resolution. If False the morph's resolution
+            will be used. If tuple the voxel size must be given in float values
+            in mm. E.g. mri_resolution=(3., 3., 3.). The default is
+            mri_resolution=False.
+            WARNING: if you have many time points the file produced can be
+            huge.
+        mri_space : bool
+            Whether the image to world registration should be in MRI space. The
+            default is mri_space=True.
+        apply_morph : bool
+            Whether to apply the precomputed morph to stc or not. The default
+            is apply_morph=False.
+        format : str
+            Either 'nifti1' (default) or 'nifti2'
+
+        Returns
+        -------
+        img : Nifti1Image | Nifti2Image
+            The image object.
+        """
+        if format != 'nifti1' and format != 'nifti2':
+            raise ValueError("invalid format specifier %s. Must be 'nifti1' or"
+                             " 'nifti2'" % format)
+        if apply_morph:
+            stc = self.__call__(stc)  # apply morph if desired
+        return _stc_as_volume(self, stc, fname=fname,
+                              mri_resolution=mri_resolution,
+                              mri_space=mri_space,
+                              format=format)
+
+    def _update_morph_data(self, data=None, kind=None):
+        """Update morph data and kind."""
+        if data is not None:
+            self.params.update(data)
```

Comment:
@TommyClausner I am not sure I get the logic here.

In practice you never call _update_morph_data with data != None and kind can be None even though you pass data that contains dict_keys(['0', '1', 'hemis', 'morph_mat', 'vertno']) so suggesting it is a surface.

can you help me understand? thanks

Reply:
I think we should have:
```
compute_source_morph
```
that does all the computations and returns
```
SourceMorph
```
object that can be used for actually applying or saving. Then `read_source_morph` will also construct the `SourceMorph` object. Then things can actually be standard class attributes (like we have everywhere else in MNE) rather than living in the `params` dict.

Reply:
hm... at some point I had mixed tested (still have the files if you wanna have it ;) ), which made it somehow necessary to update ``kind`` and ``params`` independently if I remember correctly - guess that's why it's so. Further it's used in ``morph_data_precomputed`` - but you could infer it also from the params, that's true...

So guess you're right, it's not inherently needed

Reply:
@larsoner I personally like the ``__init__`` way because of the OO taste. However I guess I see what you mean: the user could be more used to getting an object as a result of an operation, rather then initializing objects themselves, right?

I'm not sure about the attributes, because between surf and vol there is quite a difference in what you need and how many attributes this would yield. So I think having separated class and morph related "data" makes sense to me... But again, just taste, no real argument ;)